### PR TITLE
Ensure notifications for all options changes in set defaults

### PR DESCRIPTION
**Pull request by ajreynol**
This ensures that all changes to options in SetDefaults are notified in a uniform way.

The PR makes minor changes to defaults (e.g. the UF symmetry breaker is always auto-configured, but rarely is true, thus its default value is changed).  It also changes the logic slightly concerning when we notify when options changes.  We now notify whenever an option is changed (at verbosity=1), regardless of whether the change conflicts with a user option change.

This helps document the auto-configuration behavior of cvc5 for the user.

### CODE REVIEW

**Review by aniemetz — CHANGES_REQUESTED** [review]
We should introduce a wrapper to notify and modify the option, so that typos can be avoided and it is ensured that when we change the value of an option, notify is called.

**Comment by ajreynol** [comment]
This would have to be further auto-generated code in the options object I believe, since we are calling the auto-generated methods to modify the options.  Do you think we should do this?

**Comment by aniemetz** [comment]
Do we only want to notify in set_defaults?

**Comment by ajreynol** [comment]
Yes.  

The intention is for all internal changes to options to occur in `set_defaults.cpp`, and for the output to reflect only what the set defaults module is doing.  User changes (e.g. via set-option) should not be notified.

File: src/smt/set_defaults.cpp [hunk 1]
```diff
@@ -46,6 +46,56 @@ using namespace cvc5::internal::theory;
...
+  {                                                    \
+    notifyModifyOption(#optName, #value, reason);      \
+    opts.write##domain().optName = value;              \
+  }
+/**
+ * Set domain.optName to value due to reason. Notify if value changes.
+ * Print the value using a stringstream conversion.
+ *
+ * Note this macro should be used if the value passed to the macro is not
+ * concrete (e.g. is stored in a variable).
```
Comment (aniemetz) [inline src/smt/set_defaults.cpp:63]:
```suggestion
 * concrete (i.e., stored in a variable).
```

File: src/smt/set_defaults.cpp [hunk 2]
```diff
@@ -46,6 +46,56 @@ using namespace cvc5::internal::theory;
 namespace cvc5::internal {
 namespace smt {
 
+/**
+ * Set domain.optName to value due to reason. Notify if value changes.
```
Comment (aniemetz) [inline src/smt/set_defaults.cpp:50]:
```suggestion
 * Set domain.optName to value due to reason. Notify if value changes.
 * Note this macro should be used if the value is concrete.
```

File: src/smt/set_defaults.cpp [hunk 3]
```diff
@@ -46,6 +46,56 @@ using namespace cvc5::internal::theory;
...
+ */
+#define SET_AND_NOTIFY(domain, optName, value, reason) \
+  if (opts.write##domain().optName != value)           \
+  {                                                    \
+    notifyModifyOption(#optName, #value, reason);      \
+    opts.write##domain().optName = value;              \
+  }
+/**
+ * Set domain.optName to value due to reason. Notify if value changes.
+ * Print the value using a stringstream conversion.
```
Comment (aniemetz) [inline src/smt/set_defaults.cpp:60]: I think you can remove this line.

File: src/smt/set_defaults.cpp [hunk 4]
```diff
@@ -46,6 +46,56 @@ using namespace cvc5::internal::theory;
...
+  if (opts.write##domain().optName != value)                   \
+  {                                                            \
+    std::stringstream sstmp;                                   \
+    sstmp << value;                                            \
+    notifyModifyOption(#optName, sstmp.str(), reason);         \
+    opts.write##domain().optName = value;                      \
+  }
+/**
+ * Set domain.optName to value due to reason if the option was not already set
+ * by the user. Notify if value changes.
```
Comment (aniemetz) [inline src/smt/set_defaults.cpp:75]:
```suggestion
 * by the user. Notify if value changes.
 * Note this macro should be used if the value is concrete.
```

File: src/smt/set_defaults.cpp [hunk 5]
```diff
@@ -46,6 +46,56 @@ using namespace cvc5::internal::theory;
...
+  if (!opts.write##domain().optName##WasSetByUser                  \
+      && opts.write##domain().optName != value)                    \
+  {                                                                \
+    notifyModifyOption(#optName, #value, reason);                  \
+    opts.write##domain().optName = value;                          \
+  }
+/**
+ * Set domain.optName to value due to reason if the option was not already set
+ * by the user. Notify if value changes. Print the value using a stringstream
+ * conversion.
```
Comment (aniemetz) [inline src/smt/set_defaults.cpp:87]: I think the last sentence can be removed. Add a note similarly to the one for `SET_AND_NOTIFY_VAL_SYM`.

**Review by aniemetz — APPROVED** [review]
Some minor things, else LGTM.

**Pull request merged**